File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque
input = sys.stdin.readline

def dfs(graph, v, visited):
    visited[v] = True
    print(v,end=' ')
    for i in graph[v]:
        if not visited[i]:
            dfs(graph, i, visited)
def bfs(graph, v, visited):
    q = deque([v])
    while q:
        now = q.popleft()
        if not visited[now]:
            visited[now] = True
            print(now,end=' ')
            for i in graph[now]:
                q.append(i)

n, m, v = map(int, input().split())

graph = [[] for _ in range(n+1)]
visited = [False] * (n+1)

for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
for i in range(n+1):
    graph[i] = sorted(graph[i])

dfs(graph, v, visited)
visited = [False] * (n+1)
print()
bfs(graph, v, visited)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.IOException;
import java.util.*;

public class Main {
    List<List<Integer>> graph;
    boolean[] visited;
    short cnt;
    StringBuilder sb;

    public static void main(String[] args) throws IOException {
        new Main().solution();
    }

    private void solution() throws IOException {
        sb = new StringBuilder();
        graph = new ArrayList<>();
        int n = readInt(), m = readInt(), r = readInt();
        visited = new boolean[n + 1];
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i < m; i++) {
            int u = readInt(), v = readInt();
            graph.get(u).add(v);
            graph.get(v).add(u);
        }
        graph.forEach(Collections::sort);

        dfs(r);
        sb.append('\n');

        cnt = 0;
        visited = new boolean[n + 1];

        bfs(r);

        System.out.println(sb);
    }

    void dfs(int r) {
        visited[r] = true;
        sb.append(r).append(' ');
        for (int i = 0; i < graph.get(r).size(); i++) {
            int v = graph.get(r).get(i);
            if (!visited[v]) {
                dfs(v);
            }
        }
    }

    void bfs(int r) {
        Queue<Integer> q = new LinkedList<>();
        q.offer(r);
        visited[r] = true;
        sb.append(r).append(' ');
        while (!q.isEmpty()) {
            int c = q.poll();
            for (int i = 0; i < graph.get(c).size(); i++) {
                int v = graph.get(c).get(i);
                if (!visited[v]) {
                    q.offer(v);
                    visited[v] = true;
                    sb.append(v).append(' ');
                }
            }
        }
    }

    int readInt() throws IOException {
        int v, t = 0;
        while ((v = System.in.read()) > 47) {
            t = t * 10 + (v - 48);
        }
        if (v == 13) {
            System.in.read();
        }
        return t;
    }
}